### PR TITLE
Generalize XGBoostModel to GBDTModel (support LightGBM)

### LINKED ISSUE #2205: SklearnModel.model_instance should be SklearnModel.model
At present to get the underlying model in a keras model we access the `.model` attribute. For `SklearnModel` though, the underlying model instance is stored at the `.model_instance` attribute. For a uniform API, we should consider changing this attribute to `.model` and documenting it appropriately.

### PR DESCRIPTION
**Pull request by nd-02110114**
Fixed #2205

## What I did

- Generalize XGBoostModel to GBDTModel 
  - support not only XGBoost but also LightGBM
  - remove hyper parameter tuning in `fit` method
    - I think basically just tuning the learning rate and iteration is enough
  - add `fit_with_eval` method
    - sometimes, user want to hand over the train and valid data explicitly
- Fix #2205 
  - change `model_instance` to `model`

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
This is really awesome!! Only one request here. Could you add save/reload tests for `GBDTModel` as done in #2151? I'm starting a push to make sure that all DeepChem models have tests to check correctness of saving/reloading

**Comment by nd-02110114** [comment]
I added the reload test. I think it is better to write the reload test in each model test file

File: deepchem/models/tests/test_gbdt_model.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      n_estimators=50, random_state=123, verbose=False)
+  model = dc.models.GBDTModel(xgb_model, **params)
+  # fit trained model
+  model.fit(train_dataset)
+  model.save()
+  # eval model on test
+  scores = model.evaluate(test_dataset, [regression_metric])
+  assert scores[regression_metric.name] < 55
+
+  # lightgbm test
```
Comment (rbharath) [inline deepchem/models/tests/test_gbdt_model.py:43]: We should split `lightgbm` out into a separate test case here for simplicity

File: deepchem/models/tests/test_gbdt_model.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  model = dc.models.SingletaskToMultitask(tasks, xgboost_builder)
+  # fit trained model
+  model.fit(train_dataset)
+  model.save()
+  # eval model on test
+  scores = model.evaluate(test_dataset, [regression_metric])
+  score = scores[regression_metric.name]
+  assert score < 55
+
+  # lightgbm test
```
Comment (rbharath) [inline deepchem/models/tests/test_gbdt_model.py:89]: Let's split this out as a separate test too for simplicity

File: deepchem/models/tests/test_gbdt_model.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  xgb_model = xgboost.XGBClassifier(n_estimators=50, seed=123, verbose=False)
+  model = dc.models.GBDTModel(xgb_model, **params)
+  # fit trained model
+  model.fit(train_dataset)
+  model.save()
+  # eval model on test
+  scores = model.evaluate(test_dataset, [classification_metric])
+  assert scores[classification_metric.name] > .9
+
+  # lightgbm test
```
Comment (rbharath) [inline deepchem/models/tests/test_gbdt_model.py:131]: Here too

File: deepchem/models/tests/test_gbdt_model.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  reloaded_model.reload()
+  # check predictions match on test dataset
+  original_pred = model.predict(test_dataset)
+  reload_pred = reloaded_model.predict(test_dataset)
+  assert np.all(original_pred == reload_pred)
+  # eval model on test
+  scores = reloaded_model.evaluate(test_dataset, [regression_metric])
+  assert scores[regression_metric.name] < 55
+
+  # lightgbm test
```
Comment (rbharath) [inline deepchem/models/tests/test_gbdt_model.py:177]: Here too

File: docs/requirements.rst [hunk 5]
```diff
@@ -135,7 +131,5 @@ DeepChem has a number of "soft" requirements.
 .. _`RDKit`: http://www.rdkit.org/docs/Install.html
 .. _`simdna`: https://github.com/kundajelab/simdna
 .. _`Tensorflow Probability`: https://www.tensorflow.org/probability
-.. _`XGBoost`: https://xgboost.readthedocs.io/en/latest/
```
Comment (rbharath) [inline docs/requirements.rst:138]: Should we add lightgbm/xgboost as soft requirements here? Since you'd need one or the other to use the model

**Review by rbharath — APPROVED** [review]
This is looking really good! Some minor comments on testing structure but probably good to merge now even and address in subsequent PRs.

I'm going to ahead and approve this so feel free to merge when ready!

**Comment by nd-02110114** [comment]
I fixed the test structure by your comments, and the CI passed. So, I merge in!
